Initial work to respond to a DNS query

rosatolen opened the pull request:
:dancer:

Commit:
R & Y | Use unprivileged port that is not commonly used by applications

Commit:
R & Y | Initial work for responding to a query and returning an answer

Commit:
R | Use string instead of []uint8

chelseakomlo commented:
It looks like a lot of files are deleted- is there a reason for this?

chelseakomlo commented:
Specifically, why are fields, requests, and requests_test deleted?

chelseakomlo commented on nameserver/nameserver.go:
```diff
@@ -0,0 +1,33 @@
+}
+
+// Start will begin listening for DNS queries on port 8853
+func Start() error {
```
We used to have tests for this functionality, that is what the mock udp connection object is used for.

yakiradixon replied:
The mock was being used in tests that we removed, so we removed the mock as well. I explained why we removed those tests in my previous comment.

chelseakomlo replied:
Testing logic around how we are reading from a connection and doing error handling is valuable.

chelseakomlo commented on message/message.go:
```diff
@@ -0,0 +1,105 @@
+
+type header struct {
+	ID      uint16
+	QR      byte
```
If this is a byte for outside users (i.e, for anyone who is interpreting and using this information, should we really leak information about how to extract bits from a byte all throughout the project?

yakiradixon replied:
What? I don't understand the question.

chelseakomlo replied:
Whichever caller that uses headers will need to know how to handle bytes. Is there a reason why we don't use  integers?

yakiradixon replied:
We are using integers. uint8 (byte) and uint16 integer types.

yakiradixon replied:
Now that we're looking at this again we realize that we could be consistent about the names we use. byte is an alias for uint8.

rosatolen replied:
Who would need to know details outside the Message object about the header? 

Also, bytes are unsigned integers.

chelseakomlo commented on dnsserv_test.go:
```diff
@@ -1,40 +0,0 @@
-
-var _ = Suite(&DNSNameserverSuite{})
-
-func (s *DNSNameserverSuite) TestUDPConnection(c *C) {
```
Why do we remove these tests?

yakiradixon replied:
The first test was testing a function that we didn't implement ourselves (net.SplitHostPort( )). And we don't call this function anymore. We could write a test assert on what getUDPAddr( ) returns but how much value is there in asserting that the server will listen on a specific address and port?

The last two tests were asserting on the values returned by a function that isn't defined anymore with this commit.

chelseakomlo replied:
The output under test is not line 18 below, it is line 17. SplitHostPort is just a test helper.

chelseakomlo replied:
There is value in testing that we are listening to the correct address and port, right? That information is pretty important to validate...

yakiradixon replied:
Eh. The server could listen on any address and port, and I would rather focus on tests that drive out the work needed for the nameserver to respond to queries. #9 is the issue we are working on.

chelseakomlo replied:
#9 relies on the nameserver connecting to and listening to a particular address and port, no?

chelseakomlo replied:
The way this test is constructed can definitely be changed, but we should test logic around connecting, listening, and responding to errors

rosatolen replied:
Since we have changed the port 3 times in the last 3 days and it can be different for each of us depending on what we are running on 8888 or some other number. But yeah, we can implement a test for this.

Instead of focusing on creating unit tests around this functionality, we will get more value from sending messages to real nameservers and seeing their responses. After we can do that, we will be able to make unit tests. Right now, we are not providing value by testing that a function can parse the bytes from "TEST CONNECTION" because that functionality is already provided by GoLang.

My point is that we don't know enough about what the bytes look like to create a rigid unit test for them.

chelseakomlo replied:
There are different styles of implementation. Also, we do know what the bytes will look like, right? I'm confused why they would be different than what is specified in the DNS spec. 

I don't mind quick spikes to test connecting, but not covering actual implementation code with tests as we go along isn't such a good idea :/

olabini replied:
I completely agree with Y and R here - having a test that just checks the ip and port is very marginal value - and the time spent arguing on this could by itself have been spent better, I think. It is definitely a good idea not testing things where we are not responsible for the code. Further - adding unit tests has a cost, not just a benefit.

chelseakomlo commented on message/message.go:
```diff
@@ -0,0 +1,105 @@
+	}
+}
+
+func domainToLabels(domain string) []label {
```
It doesn't seem like this functionality is tested

rosatolen replied:
domainToLabels() is a helper function for Query() which is used in the test we created. We need to create more tests with different domains to continue testing this.

yakiradixon commented on message/message.go:
```diff
@@ -0,0 +1,105 @@
+
+// Query creates a DNS query based on a given domain string
+func Query(d string) *Message {
+	header := &header{
```
Got carried away with some unnecessary type conversions here.

rosatolen commented:
>It looks like a lot of files are deleted- is there a reason for this? 
>Specifically, why are fields, requests, and requests_test deleted, along with tests for how we are setting/reading the connection? 

Yes. We asked ourselves why are we creating readUDPPacket with data that does not represent the domain (i.e. []byte("TEST CONNECTION")). We decided we would be better served by doing these kinds of tests after we build the domain. Also, we were confused about the test for the length of a packet with no data in it. The UDPConn.ReadFromUDP function allows you to pass in a byte slice that you instantiate with a specific length. So you don't need to create a test to limit how large the bytes are because you can specify that length.

We removed the extractBit, extractByte, extractMultipleBits functionality because we had a hard time identifying the overall outcome for doing these things. Also, why is it important to exract an ID from the query? 

Also we don't need to extract them with bit shifting, using a for loop, and casting things to integers. We can use go's slice manipulation and boolean logic:
```
bs := make([]byte, 512)
ourUDPConn.ReadFromUDP(bs)
header.ID = bs[:1] // first two bytes
header.QR = (bs[2] & uint8(128) == uint8(128) // boolean value of most significant bit in 3rd byte
```

At the same time, why do we need to do this for all the parts of the header?

We changed the structures to be more object focused. We don't currently plan to have functions or objects outside of the Message package know about the header specifics.

chelseakomlo commented:
It is fine to change tests to make them better test code that changes as we add functionality, but we shouldn't completely remove them but keep the functionality the same.

chelseakomlo commented:
The ID is sent back to the resolver in the response, the resolver can use it to match queries to responses.

chelseakomlo commented:
I think it is better that we talk before deleting functionality that isn't understood.

yakiradixon commented:
What are we not understanding Chelsea?? And we've been talking about this pull request for the past few hours.

chelseakomlo commented:
Two pieces of feedback overall. 1) tests should be in place for functionality that we are able to test, and 2) changing how headers are parsed should be in a different commit. Rosalie's point about doing something like header.QR = (bs[2] & uint8(128) == uint8(128) is good and definitely simpler, but we probably shouldn't just delete implementation without refactoring/replacing.

rosatolen commented:
>The ID is sent back to the resolver in the response, the resolver can use it to match queries to responses. 

That doesn't mean we have to extract it. :/ The [spec](https://tools.ietf.org/html/rfc1035#section-4) shows that answers are appended after a query. In fact, [Message Compression utilizes the fact that you append the answer to the query](https://tools.ietf.org/html/rfc1035#section-4.1.4). Why would we extract each part of the header and the query only to assemble the parts together in the same order?

>I think it is better that we talk before deleting functionality that isn't understood. 

What has led you to believe that we don't understand what is going on?

Like I've said before, there is no reason to test the length of a packet the way it was tested with `TestDataFromUDPConnection`. Because the Go library UDPConn.ReadFromUDP function allows you to pass in a byte slice with a specific length. In addition, we don't need to extract each part of the header because the responses are concatenated after the query.

Please address these points becaues they are the reason why I still believe the old code is no longer providing value for us.

chelseakomlo commented:
No one has said that all parts of the header need to be extracted, so looking at which ones do/don't is to be expected.

chelseakomlo commented:
Yes, tests should definitely evolve as the implementation evolves. 

So, rather than removing the tests, I would suggest making them useful, as these were written just to get us started. CoyIm has some good examples of testing mocked connections (obviously not completely the same but relevant): 

https://github.com/twstrike/coyim/blob/121adff9ba6a4e74d3bcad6128b15f89365a04bb/xmpp/xmpp_test.go
https://github.com/twstrike/coyim/blob/b9d514ff1c964f25c723d051dd9e637e40bba4e2/xmpp/tcp_test.go
https://github.com/twstrike/coyim/blob/b9d514ff1c964f25c723d051dd9e637e40bba4e2/xmpp/helpers_for_test.go

Commit:
R & Y | Remove unnecessary type cast
